File: python/openlocationcode_test.py
#  -*- coding: utf-8 -*-

# pylint: disable=redefined-builtin
from io import open
import random
import time
import unittest
from python.openlocationcode import openlocationcode as olc

# Location of test data files.
_TEST_DATA = 'test_data'


class TestValidity(unittest.TestCase):

    def setUp(self):
        self.testdata = []
        headermap = {0: 'code', 1: 'isValid', 2: 'isShort', 3: 'isFull'}
        tests_fn = _TEST_DATA + '/validityTests.csv'
        with open(tests_fn, mode='r', encoding='utf-8') as fin:
            for line in fin:
                if line.startswith('#'):
                    continue
                td = line.strip().split(',')
                assert len(td) == len(
                    headermap), 'Wrong format of testing data: {0}'.format(line)
                # all values should be booleans except the code
                for i in range(1, len(headermap)):
                    td[i] = (td[i] == 'true')
                self.testdata.append({
                    headermap[i]: v for i, v in enumerate(td)
                })

    def test_validcodes(self):
        for td in self.testdata:
            self.assertEqual(olc.isValid(td['code']), td['isValid'], td)

    def test_fullcodes(self):
        for td in self.testdata:
            self.assertEqual(olc.isFull(td['code']), td['isFull'], td)

    def test_shortcodes(self):
        for td in self.testdata:
            self.assertEqual(olc.isShort(td['code']), td['isShort'], td)


class TestShorten(unittest.TestCase):

    def setUp(self):
        self.testdata = []
        headermap = {
            0: 'fullcode',
            1: 'lat',
            2: 'lng',
            3: 'shortcode',
            4: 'testtype'
        }
        tests_fn = _TEST_DATA + '/shortCodeTests.csv'
        with open(tests_fn, mode='r', encoding='utf-8') as fin:
            for line in fin:
                if line.startswith('#'):
                    continue
                td = line.strip().split(',')
                assert len(td) == len(
                    headermap), 'Wrong format of testing data: {0}'.format(line)
                td[1] = float(td[1])
                td[2] = float(td[2])
                self.testdata.append({
                    headermap[i]: v for i, v in enumerate(td)
                })

    def test_full2short(self):
        for td in self.testdata:
            if td['testtype'] == 'B' or td['testtype'] == 'S':
                self.assertEqual(
                    td['shortcode'],
                    olc.shorten(td['fullcode'], td['lat'], td['lng']), td)
            if td['testtype'] == 'B' or td['testtype'] == 'R':
                self.assertEqual(
                    td['fullcode'],
                    olc.recoverNearest(td['shortcode'], td['lat'], td['lng']),
                    td)


class TestEncoding(unittest.TestCase):

    def setUp(self):
        self.testdata = []
        headermap = {
            0: 'lat',
            1: 'lng',
            2: 'latInt',
            3: 'lngInt',
            4: 'length',
            5: 'code'
        }
        tests_fn = _TEST_DATA + '/encoding.csv'
        with open(tests_fn, mode='r', encoding='utf-8') as fin:
            for line in fin:
                if line.startswith('#'):
                    continue
                td = line.strip().split(',')
                assert len(td) == len(
                    headermap), 'Wrong format of testing data: {0}'.format(line)
                # First two columns are floats, next three are integers.
                td[0] = float(td[0])
                td[1] = float(td[1])
                td[2] = int(td[2])
                td[3] = int(td[3])
                td[4] = int(td[4])
                self.testdata.append({
                    headermap[i]: v for i, v in enumerate(td)
                })

    def test_converting_degrees(self):
        for td in self.testdata:
            got = olc.locationToIntegers(td['lat'], td['lng'])
            # Due to floating point precision limitations, we may get values 1 less than expected.
            self.assertTrue(
                td['latInt'] - 1 <= got[0] <= td['latInt'],
                f'Latitude conversion {td["lat"]}: want {td["latInt"]} got {got[0]}'
            )
            self.assertTrue(
                td['lngInt'] - 1 <= got[1] <= td['lngInt'],
                f'Longitude conversion {td["lng"]}: want {td["lngInt"]} got {got[1]}'
            )

    def test_encoding_degrees(self):
        # Allow a small proportion of errors due to floating point.
        allowedErrorRate = 0.05
        errors = 0
        for td in self.testdata:
            got = olc.encode(td['lat'], td['lng'], td['length'])
            if got != td['code']:
                print(
                    f'olc.encode({td["lat"]}, {td["lng"]}, {td["length"]}) want {td["code"]}, got {got}'
                )
                errors += 1
        self.assertLessEqual(errors / len(self.testdata), allowedErrorRate,
                             "olc.encode error rate too high")

    def test_encoding_integers(self):
        for td in self.testdata:
            self.assertEqual(
                td['code'],
                olc.encodeIntegers(td['latInt'], td['lngInt'], td['length']))


class TestDecoding(unittest.TestCase):

    def setUp(self):
        self.testdata = []
        headermap = {
            0: 'code',
            1: 'length',
            2: 'latLo',
            3: 'lngLo',
            4: 'latHi',
            5: 'longHi'
        }
        tests_fn = _TEST_DATA + '/decoding.csv'
        with open(tests_fn, mode='r', encoding='utf-8') as fin:
            for line in fin:
                if line.startswith('#'):
                    continue
                td = line.strip().split(',')
                assert len(td) == len(
                    headermap), 'Wrong format of testing data: {0}'.format(line)
                # all values should be numbers except the code
                for i in range(1, len(headermap)):
                    td[i] = float(td[i])
                self.testdata.append({
                    headermap[i]: v for i, v in enumerate(td)
                })

    def test_decoding(self):
        precision = 10
        for td in self.testdata:
            decoded = olc.decode(td['code'])
            self.assertAlmostEqual(decoded.latitudeLo, td['latLo'], precision,
                                   td)
            self.assertAlmostEqual(decoded.longitudeLo, td['lngLo'], precision,
                                   td)
            self.assertAlmostEqual(decoded.latitudeHi, td['latHi'], precision,
                                   td)
            self.assertAlmostEqual(decoded.longitudeHi, td['longHi'], precision,
                                   td)


class Benchmark(unittest.TestCase):

    def setUp(self):
        self.testdata = []
        for i in range(0, 100000):
            dec = random.randint(0, 15)
            lat = round(random.uniform(1, 180) - 90, dec)
            lng = round(random.uniform(1, 360) - 180, dec)
            length = random.randint(2, 15)
            if length % 2 == 1:
                length = length + 1
            self.testdata.append(
                [lat, lng, length,
                 olc.encode(lat, lng, length)])

    def test_benchmark(self):
        start_micros = round(time.time() * 1e6)
        for td in self.testdata:
            olc.encode(td[0], td[1], td[2])
        duration_micros = round(time.time() * 1e6) - start_micros
        print('Encoding benchmark: %d passes, %d usec total, %.03f usec each' %
              (len(self.testdata), duration_micros,
               duration_micros / len(self.testdata)))

        start_micros = round(time.time() * 1e6)
        for td in self.testdata:
            olc.decode(td[3])
        duration_micros = round(time.time() * 1e6) - start_micros
        print('Decoding benchmark: %d passes, %d usec total, %.03f usec each' %
              (len(self.testdata), duration_micros,
               duration_micros / len(self.testdata)))


if __name__ == '__main__':
    unittest.main()

File: elixir/test/openlocationcode_test.exs
defmodule SoferBe.Places.APIs.OpenLocationCodeTest do
  use ExUnit.Case

  # --- Validity Tests ---
  describe "validity" do
    test "valid?/1 matches test data" do
      for test_case <- read_validity_csv("validityTests.csv") do
        assert OpenLocationCode.valid?(test_case.code) == test_case.is_valid,
               "Failed on code: #{test_case.code}"
      end

      assert OpenLocationCode.valid?("849VGJQF") == false
      assert OpenLocationCode.valid?(1_234_567_890) == false
    end

    test "short?/1 matches test data" do
      for test_case <- read_validity_csv("validityTests.csv") do
        assert OpenLocationCode.short?(test_case.code) == test_case.is_short,
               "Failed on code: #{test_case.code}"
      end
    end

    test "full?/1 matches test data" do
      for test_case <- read_validity_csv("validityTests.csv") do
        assert OpenLocationCode.full?(test_case.code) == test_case.is_full,
               "Failed on code: #{test_case.code}"
      end
    end
  end

  # --- Shortening and Recovery Tests ---
  describe "shortening" do
    test "shorten/3 and recover_nearest/3 match test data" do
      for test_case <- read_shortening_csv("shortCodeTests.csv") do
        # Test shortening
        if test_case.test_type in ["B", "S"] do
          assert OpenLocationCode.shorten(
                   test_case.full_code,
                   test_case.lat,
                   test_case.lng
                 ) == {:ok, test_case.short_code}
        end

        # Test recovery
        if test_case.test_type in ["B", "R"] do
          assert OpenLocationCode.recover_nearest(
                   test_case.short_code,
                   test_case.lat,
                   test_case.lng
                 ) == {:ok, test_case.full_code}
        end
      end

      assert OpenLocationCode.shorten("CJ+2VX", 1.2, 2.3) == {:error, :full_code_expected}
    end
  end

  # --- Encoding Tests ---
  describe "encoding" do
    test "encode/3 matches test data with allowed error rate" do
      data = read_encoding_csv("encoding.csv")
      allowed_error_rate = 0.05

      results =
        for test_case <- data do
          {:ok, result} =
            OpenLocationCode.encode(
              test_case.lat,
              test_case.lng,
              test_case.length
            )

          if result == test_case.code do
            true
          else
            IO.puts(
              "encode(#{test_case.lat}, #{test_case.lng}, #{test_case.length}) want #{test_case.code}, got #{result}"
            )

            false
          end
        end

      errors = Enum.count(results, &(not &1))
      error_rate = errors / length(data)

      assert error_rate <= allowed_error_rate,
             "Encode error rate too high: #{error_rate} (allowed: #{allowed_error_rate})"
    end

    test "location_to_integers/2 matches test data with precision tolerance" do
      for test_case <- read_encoding_csv("encoding.csv") do
        case OpenLocationCode.location_to_integers(test_case.lat, test_case.lng) do
          {:ok, {got_lat_int, got_lng_int}} ->
            # Due to floating point precision limitations, we may get values 1 less than expected
            assert test_case.lat_int - 1 <= got_lat_int and got_lat_int <= test_case.lat_int,
                   "Latitude conversion #{test_case.lat}: want #{test_case.lat_int} got #{got_lat_int}"

            assert test_case.lng_int - 1 <= got_lng_int and got_lng_int <= test_case.lng_int,
                   "Longitude conversion #{test_case.lng}: want #{test_case.lng_int} got #{got_lng_int}"

          _ ->
            flunk("location_to_integers should return a tuple")
        end
      end
    end

    test "encode_integers/3 matches test data exactly" do
      for test_case <- read_encoding_csv("encoding.csv") do
        assert OpenLocationCode.encode_integers(
                 test_case.lat_int,
                 test_case.lng_int,
                 test_case.length
               ) == {:ok, test_case.code},
               "Encode integers failed for: #{inspect(test_case)}"
      end
    end
  end

  # --- Decoding Tests ---
  describe "decoding" do
    test "decode/1 matches test data" do
      for test_case <- read_decoding_csv("decoding.csv") do
        {:ok, %OpenLocationCode.CodeArea{} = decoded} = OpenLocationCode.decode(test_case.code)
        precision = 10

        assert_almost_equal(decoded.latitude_lo, test_case.lat_lo, precision, test_case)
        assert_almost_equal(decoded.longitude_lo, test_case.lng_lo, precision, test_case)
        assert_almost_equal(decoded.latitude_hi, test_case.lat_hi, precision, test_case)
        assert_almost_equal(decoded.longitude_hi, test_case.lng_hi, precision, test_case)
      end

      assert OpenLocationCode.decode("CJ+2VX") == {:error, :full_code_expected}
      assert OpenLocationCode.decode(1_234_567_890) == {:error, :invalid_code}
      assert OpenLocationCode.decode("asdsa+21") == {:error, :invalid_code}
    end
  end

  # --- Private Helper Functions for Parsing CSVs ---

  defp get_path(path) do
    "../test_data/" <> path
  end

  defp read_validity_csv(path) do
    path
    |> get_path()
    |> File.stream!()
    |> Stream.filter(&(not String.starts_with?(&1, "#")))
    |> Enum.map(fn line ->
      [code, is_valid, is_short, is_full] = String.trim(line) |> String.split(",")

      %{
        code: code,
        is_valid: is_valid == "true",
        is_short: is_short == "true",
        is_full: is_full == "true"
      }
    end)
  end

  defp read_shortening_csv(path) do
    path
    |> get_path()
    |> File.stream!()
    |> Stream.filter(&(not String.starts_with?(&1, "#")))
    |> Enum.map(fn line ->
      [full_code, lat, lng, short_code, test_type] = String.trim(line) |> String.split(",")

      %{
        full_code: full_code,
        lat: String.to_float(lat),
        lng: String.to_float(lng),
        short_code: short_code,
        test_type: test_type
      }
    end)
  end

  defp read_encoding_csv(path) do
    path
    |> get_path()
    |> File.stream!()
    |> Enum.with_index()
    |> Stream.filter(&(not String.starts_with?(elem(&1, 0), "#")))
    |> Enum.map(fn {line, index} ->
      [lat, lng, lat_int, lng_int, length, code] = String.trim(line) |> String.split(",")

      %{
        lat: elem(Float.parse(lat), 0),
        lng: elem(Float.parse(lng), 0),
        lat_int: String.to_integer(lat_int),
        lng_int: String.to_integer(lng_int),
        length: String.to_integer(length),
        code: code,
        line: index + 1
      }
    end)
  end

  defp read_decoding_csv(path) do
    path
    |> get_path()
    |> File.stream!()
    |> Stream.filter(&(not String.starts_with?(&1, "#")))
    |> Enum.map(fn line ->
      [code, _length, lat_lo, lng_lo, lat_hi, lng_hi] = String.trim(line) |> String.split(",")

      %{
        code: code,
        lat_lo: elem(Float.parse(lat_lo), 0),
        lng_lo: elem(Float.parse(lng_lo), 0),
        lat_hi: elem(Float.parse(lat_hi), 0),
        lng_hi: elem(Float.parse(lng_hi), 0)
      }
    end)
  end

  # Custom assertion to mimic Python's assertAlmostEqual
  defp assert_almost_equal(float1, float2, precision, test_case) do
    tolerance = :math.pow(10, -precision)

    assert abs(float1 - float2) < tolerance,
           "Floats not almost equal for #{inspect(test_case)}. Got #{float1}, expected #{float2}"
  end
end
